pageserver/src/storage_sync.rs

```diff
@@ -189,25 +190,29 @@ pub struct RemoteTimelineClient {
 
     storage_impl: GenericRemoteStorage,
 }
-
+enum UploadQueue {
+    Uninitialized,
+    Initialized(UploadQueueInitialized),
+}
 /// This keeps track of queued and in-progress tasks.
-#[derive(Default)]
-struct UploadQueue {
+struct UploadQueueInitialized {
     /// Counter to assign task IDs
     task_counter: u64,
 
     /// All layer files stored in the remote storage, taking into account all
     /// in-progress and queued operations
     latest_files: HashMap<RelativePath, LayerFileMetadata>,
 
-    /// Copy of latest remote metadata file. None if the upload queue hasn't been initialized yet.
-    /// Can be ahead of what's actually stored in remote storage, if new index file upload
-    /// has been scheduled but not finished yet.
-    latest_metadata: Option<TimelineMetadata>,
+    /// Metadata stored in the remote storage, taking into account all
+    /// in-progress and queued operations.
+    /// DANGER: do not return to outside world, e.g., safekeepers.
```

Comment:
> I think the comment is rather vague. "Taking into account all in-progress and queued operations" actually means that this metadata reflects local metadata at the moment of latest call to schedule_index_upload. So sometimes it is consistent with remote one sometimes not.
> 
> To me natural way would be to call it remote_consistent_metadata and remote_consistent_lsn respectively and set them only when operation completes.

Look at `schedule_layer_file_deletion` for example. It needs to create a new `IndexPart`, and for that, it needs the a `TimelineMetadata`. If there's an in-progress metadata upload, we the `schedule_layer_file_deletion` needs to use the one that is in progress. If it took whatever is the one that's actually currently on the remote, it might overwrite the update of the in-progress metadata upload.

It took me a while to figure out that this is how it works.
I don't think it's great / particularly robust, hence the DANGER comment.

Reply:
Hmm, you're right. Do you think it is a good idea to have a follow up on that?